token.php

```diff
@@ -94,11 +97,12 @@ function get_meta( $name = false, $allow_hidden = false ) {
 	* Check if a token has expired, or will expire in the next $window seconds
 	**/
 	function is_expired( $window = 0 ) {
-		if ( ! $expires = $this->get_meta( 'expires' ) ) {
+		$expires = $this->get_meta( 'expires' );
+		if ( ! $expires ) {
 			return false; // No expires value, assume it's a permanent token
 		}
 
-		if ( '0000-00-00 00:00:00' == $expires ) {
+		if ( '0000-00-00 00:00:00' === $expires ) {
```

Comment:
This I am slightly worried about. The following are all true:

`true == '0000-00-00 00:00:00'`
`0 == '0000-00-00 00:00:00'`
`00000000000000' == '0000-00-00 00:00:00'`

So if users have their expiry token set to these values then the token will be treated as if they don't expire currently, but with this change we'll break that. I have broken this before by changing the line below, so I'm being super cautious! (D33702-code)

Reply:
I agree, there is a potential logic change here.

If `$expires` is set to `0` or `00000000000000`, it'll be caught by the `! $expires` check before this, so the result doesn't change for those examples.

If it's set to `true`, it'll now skip that check, and be caught by the next one, since `(int) true === 1`, and that'll always be lower than `time()`, so `is_expired()` will now return `true` if `$expires` is set to `true`.

Looking at where `$expires` comes from, the services that use it produce it in one of two ways:

Google, Fitbit, Moves: Set it as `time() + $return['expires_in']`. The type of `$return['expires_in']` doesn't matter, since PHP will always convert it to a numeric type when adding to the integer return value of `time()`. The resulting value will always be `integer` or `double`.

Strava: Set it as the value of `$return->expires_at`. Per the [Strava docs](https://developers.strava.com/docs/authentication/#token-exchange), `expires_at` will always be an integer.

I don't have examples of other services, but given how the value is generally produced and used, it seems unlikely that they would set it to `true`.

Finally, given that Keyring 2.0 explicitly requires `expires` to be Unix timestamps, I'm inclined to think that we don't need to worry about values that aren't valid.

Reply:
@pento, this reasoning makes sense to me. We could still be cautious and move this one change to a separate diff as @scruffian suggests, but if we are aware of this potential change in the logic here, if issues with token expiry did arise it is probably just as easy to do a patch to change just this comparison back rather than needing to do a full revert.

Reply:
In the interest of keeping this PR moving, I've removed the strict comparison here, I'll open a new PR to look at implementing it.